pallets/xyk/src/lib.rs

```diff
@@ -2369,10 +2654,193 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 					bought_asset_amount,
 					min_amount_out,
 				));
+				return Ok(Default::default())
 			}
 		}
 
-		Ok(())
+		Ok(bought_asset_amount)
+	}
+
+	fn multiswap_sell_asset(
+		sender: T::AccountId,
+		swap_token_list: Vec<TokenId>,
+		sold_asset_amount: Self::Balance,
+		min_amount_out: Self::Balance,
+		err_upon_bad_slippage: bool,
+		err_upon_non_slippage_fail: bool,
+	) -> Result<Balance, DispatchError> {
+		let (
+			fee_swap_buy_and_burn_amount,
+			fee_swap_treasury_amount,
+			fee_swap_pool_fee_amount,
+			fee_swap_input_reserve,
+			fee_swap_output_reserve,
+			fee_swap_sold_asset_id,
+			fee_swap_bought_asset_id,
+		) = <Pallet<T> as PreValidateSwaps>::pre_validate_multiswap_sell_asset(
+			&sender,
+			swap_token_list.clone(),
+			sold_asset_amount,
+			min_amount_out,
+		)?;
+
+		// First execute all atomic swaps in a storage layer
+		// And then the finally bought amount is compared
+		// The bool in error represents if the fail is due to bad final slippage
+		let res_multiswap =
+			frame_support::storage::with_storage_layer(|| -> Result<Balance, DispatchError> {
```

Comment:
code style comment:

Let's move the whole functionality of `let res_multiswap =` calculation to a separate method.